velox/functions/prestosql/ArrayDuplicates.cpp

```diff
@@ -0,0 +1,193 @@
+  ///
+  /// In order to prevent copies of array elements, the function reuses the
+  /// internal elements() vector from the original ArrayVector.
+  ///
+  /// First a new vector is created containing the indices of the elements
+  /// which will be present in the output, and wrapped into a DictionaryVector.
+  /// Next the `lengths` and `offsets` vectors that control where output arrays
+  /// start and end are wrapped into the output ArrayVector.
+
+  ArrayDuplicatesFunction() {}
+
+  // Validate element type inside the vector
+  void validateElementType(const TypeKind& kind) const {
+    VELOX_USER_CHECK(
+        kind == TypeKind::TINYINT || kind == TypeKind::SMALLINT ||
+            kind == TypeKind::INTEGER || kind == TypeKind::BIGINT ||
+            kind == TypeKind::VARCHAR,
+        "array_duplicates only takes argument of type array(varchar) or array(bigint)");
+  }
+
+  // Execute function.
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      exec::Expr* caller,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // Acquire the array elements vector.
+    auto arrayVector = args.front()->as<ArrayVector>();
+    auto elementsVector = arrayVector->elements();
+    validateElementType(elementsVector->typeKind());
+
+    auto elementsRows =
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+
+    vector_size_t elementsCount = elementsRows.size();
+    vector_size_t rowCount = arrayVector->size();
+
+    // Allocate new vectors for indices, length and offsets.
+    memory::MemoryPool* pool = context->pool();
+    BufferPtr newIndices =
+        AlignedBuffer::allocate<vector_size_t>(elementsCount, pool);
+    BufferPtr newLengths =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+    BufferPtr newOffsets =
+        AlignedBuffer::allocate<vector_size_t>(rowCount, pool);
+
+    // Pointers and cursors to the raw data.
+    vector_size_t indicesCursor = 0;
+    auto* rawNewIndices = newIndices->asMutable<vector_size_t>();
+    auto* rawSizes = newLengths->asMutable<vector_size_t>();
+    auto* rawOffsets = newOffsets->asMutable<vector_size_t>();
+
+    // Process the rows: use a hashmap to store unique values and
+    // whether it occurred once or more than once.
+    folly::F14FastMap<T, bool> uniqueMap;
+
+    rows.applyToSelected([&](vector_size_t row) {
+      auto size = arrayVector->sizeAt(row);
+      auto offset = arrayVector->offsetAt(row);
+
+      rawOffsets[row] = indicesCursor;
+      vector_size_t numOfNulls = 0;
+      for (vector_size_t i = offset; i < offset + size; ++i) {
+        if (elements->isNullAt(i)) {
+          numOfNulls++;
+          if (numOfNulls == 2) {
+            rawNewIndices[indicesCursor++] = i;
+          }
+        } else {
+          auto value = elements->valueAt<T>(i);
+          if (uniqueMap.contains(value)) {
```

Comment:
I still feel we should use F14Map here instead of `F14Set.insert(value).second` because we need to find all "duplicates" instead of "unique" values. The bool flag in the Map is to indicate whether the value occurs once or more than once.
cc @mbasmanova

Reply:
I think I understand this. Sounds fine. Thanks for clarifying.

Reply:
nit: you may want to reduce the number of lookups in the map by using uniqueMap.find() which returns an iterator to the entry and can be used to check whether the entry exists as well as what is the value.